File: backend/manish-hardware-backend/src/main/java/com/manish/hardware/service/ProductService.java
package com.manish.hardware.service;

import com.manish.hardware.model.Product;
import com.manish.hardware.model.ProductTranslation;
import com.manish.hardware.repository.ProductRepository;
import com.manish.hardware.repository.ProductTranslationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private ProductTranslationRepository translationRepository;

    // Get all active products
    public List<Product> getAllActiveProducts() {
        return productRepository.findByIsActiveTrue();
    }

    // Get all active products with pagination
    public Page<Product> getAllActiveProducts(Pageable pageable) {
        return productRepository.findByIsActiveTrue(pageable);
    }

    // Get products by category
    public List<Product> getProductsByCategory(Long categoryId) {
        return productRepository.findByCategoryIdAndIsActiveTrue(categoryId);
    }

    // Get products by category with pagination
    public Page<Product> getProductsByCategory(Long categoryId, Pageable pageable) {
        return productRepository.findByCategoryIdAndIsActiveTrue(categoryId, pageable);
    }

    // Get product by ID
    public Optional<Product> getProductById(Long id) {
        return productRepository.findById(id);
    }

    // Get product by ID with translations
    public Optional<Product> getProductByIdWithTranslations(Long id) {
        return productRepository.findByIdWithTranslations(id);
    }

    // Get product by ID and language
    public Optional<Product> getProductByIdAndLanguage(Long id, String languageCode) {
        return productRepository.findByIdAndLanguageCode(id, languageCode);
    }

    // Search products by name
    public Page<Product> searchProducts(String searchTerm, Pageable pageable) {
        return productRepository.searchByName(searchTerm, pageable);
    }

    // Search products by name within category
    public Page<Product> searchProducts(String searchTerm, Long categoryId, Pageable pageable) {
        return productRepository.searchByNameAndCategory(searchTerm, categoryId, pageable);
    }

    // Create product
    @Transactional
    public Product createProduct(Product product) {
        return productRepository.save(product);
    }

    // Update product
    @Transactional
    public Product updateProduct(Long id, Product productDetails) {
        Optional<Product> optionalProduct = productRepository.findById(id);
        if (optionalProduct.isPresent()) {
            Product product = optionalProduct.get();
            product.setCategoryId(productDetails.getCategoryId());
            product.setBrand(productDetails.getBrand());
            product.setImageUrl(productDetails.getImageUrl());
            product.setTechnicalSpecs(productDetails.getTechnicalSpecs());
            product.setUsageInfo(productDetails.getUsageInfo());
            product.setIsActive(productDetails.getIsActive());
            return productRepository.save(product);
        }
        return null;
    }

    // Delete product (soft delete)
    @Transactional
    public boolean deleteProduct(Long id) {
        Optional<Product> optionalProduct = productRepository.findById(id);
        if (optionalProduct.isPresent()) {
            Product product = optionalProduct.get();
            product.setIsActive(false);
            productRepository.save(product);
            return true;
        }
        return false;
    }

    // Hard delete product
    @Transactional
    public boolean hardDeleteProduct(Long id) {
        if (productRepository.existsById(id)) {
            translationRepository.deleteByProductId(id);
            productRepository.deleteById(id);
            return true;
        }
        return false;
    }

    // Add translation to product
    @Transactional
    public ProductTranslation addTranslation(Long productId, ProductTranslation translation) {
        translation.setProductId(productId);
        return translationRepository.save(translation);
    }

    // Update translation
    @Transactional
    public ProductTranslation updateTranslation(Long productId, String languageCode,
            ProductTranslation translationDetails) {
        Optional<ProductTranslation> optionalTranslation = translationRepository
                .findByProductIdAndLanguageCode(productId, languageCode);
        if (optionalTranslation.isPresent()) {
            ProductTranslation translation = optionalTranslation.get();
            translation.setName(translationDetails.getName());
            translation.setDescription(translationDetails.getDescription());
            return translationRepository.save(translation);
        }
        return null;
    }

    // Delete translation
    @Transactional
    public boolean deleteTranslation(Long productId, String languageCode) {
        Optional<ProductTranslation> optionalTranslation = translationRepository
                .findByProductIdAndLanguageCode(productId, languageCode);
        if (optionalTranslation.isPresent()) {
            translationRepository.delete(optionalTranslation.get());
            return true;
        }
        return false;
    }

    // Get translations for a product
    public List<ProductTranslation> getTranslations(Long productId) {
        return translationRepository.findByProductId(productId);
    }

    // Count products in category
    public long countProductsInCategory(Long categoryId) {
        return productRepository.countByCategoryIdAndIsActiveTrue(categoryId);
    }

    // Get translations by language code
    public List<ProductTranslation> getTranslationsByLanguageCode(String languageCode) {
        return translationRepository.findByLanguageCode(languageCode);
    }

    // Check if translation exists
    public boolean translationExists(Long productId, String languageCode) {
        return translationRepository.existsByProductIdAndLanguageCode(productId, languageCode);
    }

    // Get count of all active products
    public long getActiveProductCount() {
        return productRepository.countByIsActiveTrue();
    }
}

File: mobile/src/services/api/productService.ts
import apiClient from './apiClient';
import { PaginatedResponse } from './types';

export interface ProductTranslation {
  id: number;
  productId: number;
  languageCode: string;
  name: string;
  description: string;
}

export interface Product {
  id: number;
  categoryId: number;
  brand: string;
  imageUrl: string;
  technicalSpecs: string;
  usageInfo: string;
  isActive: boolean;
  translations?: ProductTranslation[];
  createdAt?: string;
  updatedAt?: string;
}

export interface ProductQueryParams {
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
}

export const productService = {
  // Get all products with pagination
  getAllProducts: async (params?: ProductQueryParams): Promise<PaginatedResponse<Product>> => {
    const response = await apiClient.get<PaginatedResponse<Product>>('/products', { params });
    return response.data;
  },

  // Get products by category with pagination
  getProductsByCategory: async (
    categoryId: number,
    params?: ProductQueryParams
  ): Promise<PaginatedResponse<Product>> => {
    const response = await apiClient.get<PaginatedResponse<Product>>(
      `/products/category/${categoryId}`,
      { params }
    );
    return response.data;
  },

  // Get product by ID with translations
  getProductById: async (id: number): Promise<Product> => {
    const response = await apiClient.get<Product>(`/products/${id}`);
    return response.data;
  },

  // Get product by ID and language
  getProductByIdAndLanguage: async (id: number, languageCode: string): Promise<Product> => {
    const response = await apiClient.get<Product>(`/products/${id}/lang/${languageCode}`);
    return response.data;
  },

  // Search products
  searchProducts: async (
    query: string,
    categoryId?: number,
    params?: ProductQueryParams
  ): Promise<PaginatedResponse<Product>> => {
    const response = await apiClient.get<PaginatedResponse<Product>>('/products/search', {
      params: {
        q: query,
        categoryId,
        ...params,
      },
    });
    return response.data;
  },

  // Get product translations
  getProductTranslations: async (productId: number): Promise<ProductTranslation[]> => {
    const response = await apiClient.get<ProductTranslation[]>(
      `/products/${productId}/translations`
    );
    return response.data;
  },

  // Admin: Create product
  createProduct: async (product: Partial<Product>): Promise<Product> => {
    const response = await apiClient.post<Product>('/products', product);
    return response.data;
  },

  // Admin: Update product
  updateProduct: async (id: number, product: Partial<Product>): Promise<Product> => {
    const response = await apiClient.put<Product>(`/products/${id}`, product);
    return response.data;
  },

  // Admin: Delete product (soft delete)
  deleteProduct: async (id: number): Promise<void> => {
    await apiClient.delete(`/products/${id}`);
  },

  // Admin: Add translation to product
  addTranslation: async (
    productId: number,
    translation: Partial<ProductTranslation>
  ): Promise<ProductTranslation> => {
    const response = await apiClient.post<ProductTranslation>(
      `/products/${productId}/translations`,
      translation
    );
    return response.data;
  },

  // Admin: Update translation
  updateTranslation: async (
    productId: number,
    languageCode: string,
    translation: Partial<ProductTranslation>
  ): Promise<ProductTranslation> => {
    const response = await apiClient.put<ProductTranslation>(
      `/products/${productId}/translations/${languageCode}`,
      translation
    );
    return response.data;
  },

  // Admin: Delete translation
  deleteTranslation: async (productId: number, languageCode: string): Promise<void> => {
    await apiClient.delete(`/products/${productId}/translations/${languageCode}`);
  },

  // Get translations by language code
  getTranslationsByLanguage: async (languageCode: string): Promise<ProductTranslation[]> => {
    const response = await apiClient.get<ProductTranslation[]>(`/products/translations/language/${languageCode}`);
    return response.data;
  },

  // Check if translation exists
  translationExists: async (productId: number, languageCode: string): Promise<boolean> => {
    const response = await apiClient.get<boolean>(`/products/${productId}/translations/${languageCode}/exists`);
    return response.data;
  },

  // Count products in category
  countInCategory: async (categoryId: number): Promise<number> => {
    const response = await apiClient.get<number>(`/products/count/category/${categoryId}`);
    return response.data;
  },

  // Admin: Hard delete product
  hardDeleteProduct: async (id: number): Promise<void> => {
    await apiClient.delete(`/products/${id}/hard`);
  },
};
